File: platform/polyglot/raise_exceptions/py/src/common_exceptions/fastapi/normalizers.py
"""
Validation error normalizers for FastAPI/Pydantic.

Converts Pydantic validation errors to standardized format.
"""

from typing import Any, Dict, List

from ..logger import create

logger = create("common_exceptions", __file__)


def normalize_pydantic_errors(errors: List[Dict[str, Any]]) -> List[Dict[str, str]]:
    """
    Normalize Pydantic validation errors to standardized format.

    Pydantic errors have shape:
        {"loc": ["body", "user", "email"], "msg": "...", "type": "..."}

    Normalized output:
        {"field": "body.user.email", "message": "...", "code": "..."}

    Args:
        errors: List of Pydantic error dicts

    Returns:
        List of normalized error dicts
    """
    normalized = []

    for error in errors:
        # Convert loc tuple/list to dot-notation path
        loc = error.get("loc", [])
        field_path = ".".join(str(part) for part in loc)

        # Extract message
        message = error.get("msg", "Invalid value")

        # Extract error type as code
        error_type = error.get("type", "")
        code = _normalize_error_type(error_type)

        normalized_error = {
            "field": field_path,
            "message": _capitalize_first(message),
        }

        if code:
            normalized_error["code"] = code

        logger.debug(f"Normalized validation error: {field_path} - {message}")
        normalized.append(normalized_error)

    return normalized


def _normalize_error_type(error_type: str) -> str:
    """
    Normalize Pydantic error type to standardized code.

    Args:
        error_type: Pydantic error type (e.g., "value_error.email")

    Returns:
        Normalized code (e.g., "format.email")
    """
    # Map common Pydantic error types to codes
    type_mapping = {
        "value_error.missing": "required",
        "value_error.email": "format.email",
        "value_error.url": "format.url",
        "value_error.uuid": "format.uuid",
        "type_error.integer": "type.integer",
        "type_error.string": "type.string",
        "type_error.bool": "type.boolean",
        "type_error.none.not_allowed": "type.not_null",
        "value_error.any_str.min_length": "string.min_length",
        "value_error.any_str.max_length": "string.max_length",
        "value_error.number.not_gt": "number.greater_than",
        "value_error.number.not_ge": "number.greater_than_or_equal",
        "value_error.number.not_lt": "number.less_than",
        "value_error.number.not_le": "number.less_than_or_equal",
        "value_error.list.min_items": "array.min_items",
        "value_error.list.max_items": "array.max_items",
        "value_error.const": "enum.invalid",
    }

    # Check exact match first
    if error_type in type_mapping:
        return type_mapping[error_type]

    # Check prefix matches
    for prefix, code in type_mapping.items():
        if error_type.startswith(prefix.split(".")[0]):
            return code

    # Default: convert underscores to dots
    return error_type.replace("_", ".")


def _capitalize_first(message: str) -> str:
    """Capitalize first letter of message."""
    if not message:
        return message
    return message[0].upper() + message[1:]


logger.debug("Pydantic normalizers initialized")

File: platform/polyglot/raise_exceptions/mjs/src/fastify/normalizers.ts
/**
 * Validation error normalizers for Fastify/AJV/Zod.
 *
 * Converts validation library errors to standardized format.
 */

import { create } from '../logger.js';

const logger = create('common-exceptions', __filename);

/**
 * Normalized field error structure.
 */
export interface NormalizedFieldError {
  field: string;
  message: string;
  code?: string;
}

/**
 * AJV error shape.
 */
export interface AjvError {
  keyword: string;
  dataPath?: string;
  instancePath?: string;
  message?: string;
  params?: Record<string, unknown>;
  schemaPath?: string;
}

/**
 * Zod error shape.
 */
export interface ZodIssue {
  code: string;
  path: (string | number)[];
  message: string;
}

/**
 * Normalize AJV validation errors to standardized format.
 *
 * AJV errors have shape:
 *   { keyword: "...", instancePath: "/user/email", message: "..." }
 *
 * Normalized output:
 *   { field: "user.email", message: "...", code: "..." }
 *
 * @param errors - Array of AJV error objects
 * @returns Array of normalized error objects
 */
export function normalizeAjvErrors(errors: AjvError[]): NormalizedFieldError[] {
  const normalized: NormalizedFieldError[] = [];

  for (const error of errors) {
    // Convert path: /user/email or instancePath to dot notation
    const path = error.instancePath || error.dataPath || '';
    const fieldPath = path
      .replace(/^\//, '') // Remove leading slash
      .replace(/\//g, '.'); // Convert slashes to dots

    // Get message
    const message = error.message || 'Invalid value';

    // Map keyword to code
    const code = normalizeAjvKeyword(error.keyword);

    const normalizedError: NormalizedFieldError = {
      field: fieldPath || 'body',
      message: capitalizeFirst(message),
    };

    if (code) {
      normalizedError.code = code;
    }

    logger.debug(`Normalized AJV error: ${normalizedError.field} - ${message}`);
    normalized.push(normalizedError);
  }

  return normalized;
}

/**
 * Normalize Zod validation errors to standardized format.
 *
 * Zod errors have shape:
 *   { code: "...", path: ["user", "email"], message: "..." }
 *
 * Normalized output:
 *   { field: "user.email", message: "...", code: "..." }
 *
 * @param issues - Array of Zod issue objects
 * @returns Array of normalized error objects
 */
export function normalizeZodErrors(issues: ZodIssue[]): NormalizedFieldError[] {
  const normalized: NormalizedFieldError[] = [];

  for (const issue of issues) {
    // Convert path array to dot notation
    const fieldPath = issue.path.map(String).join('.');

    // Get message
    const message = issue.message || 'Invalid value';

    // Map Zod code to standardized code
    const code = normalizeZodCode(issue.code);

    const normalizedError: NormalizedFieldError = {
      field: fieldPath || 'body',
      message: capitalizeFirst(message),
    };

    if (code) {
      normalizedError.code = code;
    }

    logger.debug(`Normalized Zod error: ${normalizedError.field} - ${message}`);
    normalized.push(normalizedError);
  }

  return normalized;
}

/**
 * Map AJV keyword to standardized code.
 */
function normalizeAjvKeyword(keyword: string): string | undefined {
  const keywordMapping: Record<string, string> = {
    required: 'required',
    type: 'type.invalid',
    format: 'format.invalid',
    minimum: 'number.minimum',
    maximum: 'number.maximum',
    minLength: 'string.min_length',
    maxLength: 'string.max_length',
    pattern: 'string.pattern',
    enum: 'enum.invalid',
    const: 'enum.invalid',
    minItems: 'array.min_items',
    maxItems: 'array.max_items',
    uniqueItems: 'array.unique_items',
    additionalProperties: 'object.additional_properties',
  };

  return keywordMapping[keyword] ?? keyword;
}

/**
 * Map Zod code to standardized code.
 */
function normalizeZodCode(code: string): string | undefined {
  const codeMapping: Record<string, string> = {
    invalid_type: 'type.invalid',
    invalid_literal: 'enum.invalid',
    unrecognized_keys: 'object.additional_properties',
    invalid_union: 'union.invalid',
    invalid_enum_value: 'enum.invalid',
    invalid_string: 'string.invalid',
    too_small: 'size.too_small',
    too_big: 'size.too_big',
    invalid_date: 'format.date',
    custom: 'custom',
  };

  return codeMapping[code] ?? code;
}

/**
 * Capitalize first letter of message.
 */
function capitalizeFirst(message: string): string {
  if (!message) return message;
  return message.charAt(0).toUpperCase() + message.slice(1);
}

logger.debug('Validation normalizers initialized');
